src/mpid/ch4/netmod/ucx/ucx_progress.c

```diff
@@ -9,14 +9,10 @@
 
 #include "mpidimpl.h"
 #include "ucx_impl.h"
-//#include "events.h"
 
-static int am_handler(void *msg, size_t msg_sz);
-static void handle_am_recv(void *request, ucs_status_t status, ucp_tag_recv_info_t * info);
-
-static int am_handler(void *msg, size_t msg_sz)
+static void *am_buf;
```

Comment:
What is the thread-safety for this buffer?

Reply:
There is none at the moment. The buffer is allocated/consumed/freed during `int MPIDI_UCX_progress`. I suppose we could have a buffer per vci to allow concurrent progress calls?

Reply:
I guess I was confused. Before, `am_handler` is essentially a sender-side completion call-back, right? Isn't `handle_am_recv` target-side handler call back?

Reply:
`am_handler` (`handle_am_recv` before) is a target-side callback executed by UCX when `ucp_tag_msg_recv_nb` completes.

Reply:
The difference is that `handle_am_recv` used to be a no-op, and all the actual MPICH active message handling was done by `am_handler` once the progress loop determined the request was complete. Now, UCX just executes `am_handler` directly.

Reply:
> `am_handler` (`handle_am_recv` before) is a target-side callback

So how is it possible to have sender-side malloc `am_buf` and let target-side consume it?

Reply:
Never mind, I was confused. The progress is target-side